core/src/main/java/org/apache/druid/timeline/partition/DimensionRangeShardSpec.java

```diff
@@ -223,16 +208,29 @@ public boolean possibleInDomain(Map<String, RangeSet<String>> domain)
 
       // EffectiveDomain[i] = QueryDomain[i] INTERSECTION SegmentRange[i]
       RangeSet<String> effectiveDomainForDimension = queryDomainForDimension.subRangeSet(rangeTillSegmentBoundary);
+
+      // Create an iterator to use for checking if the RangeSet is empty or is a singleton. This is significantly
+      // faster than using isEmpty() and equals(), because those methods call size() internally, which iterates
+      // the entire RangeSet.
+      final Iterator<Range<String>> effectiveDomainRangeIterator = effectiveDomainForDimension.asRanges().iterator();
+
       // Prune segment because query domain is out of segment range
-      if (effectiveDomainForDimension.isEmpty()) {
+      if (!effectiveDomainRangeIterator.hasNext()) {
         return false;
       }
 
-      // EffectiveDomain is singleton and lies only on the boundaries -> consider next dimensions
-      effectiveDomainIsStart = effectiveDomainIsStart
-                                && isRangeSetSingletonWithVal(effectiveDomainForDimension, segmentStart.get(i));
-      effectiveDomainIsEnd = effectiveDomainIsEnd
-                           && isRangeSetSingletonWithVal(effectiveDomainForDimension, segmentEnd.get(i));
+      final Range<String> firstRange = effectiveDomainRangeIterator.next();
+
+      if (!effectiveDomainRangeIterator.hasNext()) {
+        // Effective domain contained only one Range.
+        // If it's a singleton and lies only on the boundaries -> consider next dimensions
+        effectiveDomainIsStart = effectiveDomainIsStart
+                                 && segmentStart.get(i) != null
+                                 && firstRange.equals(Range.singleton(segmentStart.get(i)));
+        effectiveDomainIsEnd = effectiveDomainIsEnd
+                               && segmentEnd.get(i) != null
+                               && firstRange.equals(Range.singleton(segmentEnd.get(i)));
+      }
```

Comment:
`else {
  return true;
}`
This snippet might be required at this point since effective domains are no longer singleton, and we cannot prune further?
If not, we may go to the next dimension and prune a segment that must not be pruned.

Reply:
Sorry, missed @hqx871 's comment. It suggests equivalent changes.
As for the error: Returning true early means that we'll never reach a dimension where start > end

Reply:
@Test
  public void testPossibleInDomain_FalsePruning()
  {
    setDimensions("planet", "country", "city");

    final StringTuple start = StringTuple.create("Earth", "France", "Paris");
    final StringTuple end = StringTuple.create("Mars", "USA", "New York");

    final RangeSet<String> universalSet = TreeRangeSet.create();
    universalSet.add(Range.all());

    ShardSpec shard = new DimensionRangeShardSpec(dimensions, start, end, 0, null);
    Map<String, RangeSet<String>> domain = new HashMap<>();

    // {Earth} U (Mars, INF) * (USA, INF) * (-INF, INF)
    populateDomain(
        domain,
        getUnion(
            getRangeSet(Range.singleton("Earth")),
            getRangeSet(Range.singleton("Mars"))
        ),
        getUnion(
            getRangeSet(Range.greaterThan("USA"))
        ),
        universalSet
    );
    assertTrue(shard.possibleInDomain(domain));
  }